pandas/tests/reshape/merge/test_merge.py

```diff
@@ -1055,22 +1042,53 @@ def test_merge_incompat_infer_boolean_object(self):
         result = pd.merge(df2, df1, on='key')
         assert_frame_equal(result, expected)
 
+    @pytest.mark.parametrize('df1_vals, df2_vals, left_type, right_type', [
+
+        # infer to numeric
+        ([0, 1, 2], ["0", "1", "2"],
+         is_integer_dtype, is_object_dtype),
```

Comment:
I don't understand this first one... Why is the dtype of `A` int?

```python
In [7]: a = pd.DataFrame({"A": [1, 2], "B": [3, 4]})

In [8]: b = pd.DataFrame({"A": ['1', '2'], "B": [3, 4]})

In [9]: pd.merge(a, b, on="A").A.dtype
Out[9]: dtype('int64')
```

I thought the expected behavior was to coerce both to object when the dtypes don't match.

Reply:
we actually infer this because this is what numpy does. we could exclude non-numeric inference though, let me see